src/fd.cpp

```diff
@@ -0,0 +1,147 @@
+#ifdef _WIN32
+#include <winsock2.h>
+#endif
+#include <Rcpp.h>
+#include <unistd.h>
+#include <cstdlib>
+#include "later.h"
+#include "callback_registry_table.h"
+
+extern CallbackRegistryTable callbackRegistryTable;
+
+typedef struct ThreadArgs_s {
+  SEXP callback;
+  R_xlen_t num_fds;
+  std::unique_ptr<std::vector<int>> fds;
+  fd_set read_fds;
+  struct timeval tv;
+  int max_fd;
+  int loop;
+  bool flag;
+} ThreadArgs;
+
+static void later_callback(void *arg) {
+
+  std::unique_ptr<std::shared_ptr<ThreadArgs>> argsptr(static_cast<std::shared_ptr<ThreadArgs>*>(arg));
+  std::shared_ptr<ThreadArgs> args = *argsptr;
+
+  SEXP results, call;
+  // TODO: actually pre-allocate call on main thread, values can be safely updated from background thread
+  // only call Rf_eval() here to minimise runtime in later callback.
+  PROTECT(results = Rf_allocVector(LGLSXP, args->num_fds));
+  std::memcpy((void *) DATAPTR_RO(results), args->fds->data(), args->num_fds * sizeof(int));
+  PROTECT(call = Rf_lcons(args->callback, Rf_cons(results, R_NilValue)));
+  Rf_eval(call, R_GlobalEnv);
+  UNPROTECT(2);
+  R_ReleaseObject(args->callback);
+
+}
+
+// CONSIDER: upgrade select() to poll() / WSAPoll().
+// CONSIDER: add method for HANDLES on Windows. Assuming we only accept integer
+// values for both, we could use heuristics: check if it is a valid SOCKET or
+// else assume HANDLE - but this is not fool-proof.
+// Otherwise would require an interface for user to specify type.
+static void *select_thread(void *arg) {
+
+  std::unique_ptr<std::shared_ptr<ThreadArgs>> argsptr(static_cast<std::shared_ptr<ThreadArgs>*>(arg));
+  std::shared_ptr<ThreadArgs> args = *argsptr;
+
+  // CONSIDER: if we should be checking more than read activity, i.e. write and error.
+  // Could check all types for all fds, which would be inefficient but
+  // otherwise would require user interface to specify events for each fd
+  int ready = select(args->max_fd + 1, &args->read_fds, NULL, NULL, args->flag ? &args->tv : NULL);
```

Comment:
For POSIX at least, I think it makes a ton of sense to have the R function take `readfds`, `writefds`, and `exceptfds` arguments.

Reply:
I guess it makes sense for sockets on win32 too.

Reply:
With poll(), it's less overhead to simply register fds for both read and write events. I'll investigate a bit more - especially on Windows as the API there seems to be a bit different for exception/error events. Having three 'fd' arguments might seem a bit verbose, as I expect most use cases will center on read activity.

Reply:
The logic with `poll()` is to register `POLLIN` (for readfds) and `POLLOUT` (for writefds) or `0` for neither (exceptfds) as exceptions / errors are always returned regardless.

I have hence implemented in [daaaa43](https://github.com/shikokuchuo/later/commit/daaaa437bcf1363eb90a42024cd9a6ad1e2652a8) the `readfds`, `writefds` and `exceptfds` arguments at `later_fd()` as you suggested above. Happy to tweak the interface in case you have further ideas.